crates/storage/nippy-jar/src/lib.rs

```diff
@@ -0,0 +1,407 @@
+use serde::{Deserialize, Serialize};
+use std::{
+    fs::File,
+    io::{Read, Seek, SeekFrom, Write},
+    path::{Path, PathBuf},
+    sync::Mutex,
+};
+use thiserror::Error;
+use zstd::bulk::Decompressor;
+
+pub mod compression;
+use compression::{Compression, Compressors};
+
+const NIPPY_JAR_VERSION: usize = 1;
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct NippyJar {
+    /// Version
+    version: usize,
+    /// Number of columns
+    columns: usize,
+    /// Compressor if required
+    compressor: Option<Compressors>,
+    #[serde(skip)]
+    /// Data path for file. Index file will be `{path}.idx`
+    path: Option<PathBuf>,
+    /// soon
+    bloom_filter: bool,
+    /// soon
+    phf: bool,
+}
+
+impl NippyJar {
+    /// Creates new [`NippyJar`].
+    pub fn new(columns: usize, bloom_filter: bool, phf: bool, path: &Path) -> Self {
+        NippyJar {
+            version: NIPPY_JAR_VERSION,
+            columns,
+            bloom_filter,
+            phf,
+            compressor: None,
+            path: Some(path.to_path_buf()),
+        }
+    }
+
+    /// Adds [`compression::Zstd`] compression.
+    pub fn with_zstd(mut self, use_dict: bool, max_dict_size: usize) -> Self {
+        self.compressor =
+            Some(Compressors::Zstd(compression::Zstd::new(use_dict, max_dict_size, self.columns)));
+        self
+    }
+
+    /// Loads the file configuration and returns [`Self`].
+    pub fn load(path: &Path) -> Result<Self, NippyJarError> {
+        let mut file = File::open(path)?;
+        let mut obj: Self = bincode::deserialize_from(&mut file)?;
+
+        obj.path = Some(path.to_path_buf());
+        Ok(obj)
+    }
+
+    /// Returns the path from the data file
+    pub fn data_path(&self) -> PathBuf {
+        self.path.clone().expect("exists")
+    }
+
+    /// Returns the path from the index file
+    pub fn index_path(&self) -> PathBuf {
+        let data_path = self.data_path();
+        data_path
+            .parent()
+            .expect("exists")
+            .join(format!("{}.idx", data_path.file_name().expect("exists").to_string_lossy()))
```

Comment:
nit: feels a bit fragile, let's maybe specify the `self.path` as data dir and derive data and index file paths from it